src/abstractdataframe/sort.jl

```diff
@@ -682,16 +701,118 @@ function Base.sort!(df::AbstractDataFrame, cols=All();
                     lt::Union{Function, AbstractVector{<:Function}}=isless,
                     by::Union{Function, AbstractVector{<:Function}}=identity,
                     rev::Union{Bool, AbstractVector{Bool}}=false,
-                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward)
+                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward,
+                    checkunique::Bool=false)
 
     # exclude AbstractVector as in that case cols can contain order(...) clauses
     if cols isa MultiColumnIndex && !(cols isa AbstractVector)
         cols = index(df)[cols]
     end
     ord = ordering(df, cols, lt, by, rev, order)
     _alg = Sort.defalg(df, ord; alg=alg, cols=cols)
+    checkunique && _perform_uniqueness_checks(df, cols, lt, by, order)
     return sort!(df, _alg, ord)
 end
 
-Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm, o::Base.Sort.Ordering) =
+function Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm,
+                    o::Base.Sort.Ordering, checkunique::Bool=false)
+    if checkunique
+        is_complex(o) &&
+            throw(ArgumentError("Orderings created with lt or by set to " *
+                                "something other than isless and identity " *
+                                "respectively, along checkunique=true are not supported."))
+        allunique(df, col_idxs) ||
+            throw(ArgumentError("Non-unique elements found. Multiple orders are valid."))
+    end
     permute!(df, _sortperm(df, a, o))
+end
+
+#= Note to the reviewers:
+ForwardOrdering is the only order that we can be certain is simple, since it's 
+a trivial subtype of the abstract type order. It can't be instantiated manually
+by users with a non-trivial by or lt function since no such constructor exists 
+in Base. I realize this is tecnically counting on Base not changing it's behaviour,
+but as it stands, there is no other way to check if the order is complex, since
+it does not even have any fields we could check.
+
+As for the case with ReverseOrdering, the `fwd` field is some ordering, either
+ForwardOrdering, By or Lt. If it's ForwardOrdering, that implies the order does
+not contain any non-standard lt or by functions. In the case where `o` is either
+Lt or By, we check if the function is standard.
+=#
+is_complex(o::ForwardOrdering) = false
```

Comment:
We now do not cover all `Base.Sort.Ordering` subtypes. E.g. there is `Base.Order.Perm` that you do not cover.

However, maybe it is enough to define:
```
is_complex(o::Base.Sort.Ordering) = throw(ArgumentError("ordering type $(typeof(o)) is currently not supported with `checkunique=true`"))
```
(or something similar)

What I intend to have an informative error, but not to spend to much time trying to cover cases that probably will happen super rarely.

Reply:
Okay. I will take a look at the as of yet unsupported `Ordering` types. If I think I can do the checking relatively simply, I'll implement it. Otherwise I'll take your suggestion and let the user know it's unsupported